src/stats/cdf.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2016 Johannes Köster.
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Support for discrete probability distributions in terms of cumulative distribution
+//! functions (CDF).
+
+use std::f64;
+use std::iter;
+use std::slice;
+
+use num::traits::{cast, NumCast};
+use itertools::Itertools;
+
+use stats::logprobs::{self, LogProb};
+
+
+/// Implementation of a cumulative distribution function.
+#[derive(Debug, Clone)]
+pub struct CDF<T: PartialOrd> {
+    inner: Vec<(T, LogProb)>
+}
+
+
+impl<T: PartialOrd> CDF<T> {
+    /// Create CDF from given probability mass function (PMF). The PMF may contain duplicate values
+    /// the probabilities of which are summed during generation of the CDF.
+    ///
+    /// # Arguments
+    ///
+    /// * `pmf` - the PMF as a vector of value/probability pairs
+    pub fn from_pmf(mut entries: Vec<(T, LogProb)>) -> Self {
+        entries.sort_by(|&(ref a, _), &(ref b, _)| a.partial_cmp(b).unwrap());
+        let mut inner: Vec<(T, LogProb)> = Vec::new();
+        for mut e in entries.into_iter() {
+            let p = logprobs::add(inner.last().map_or(f64::NEG_INFINITY, |e| e.1), e.1);
+            if !inner.is_empty() && inner.last().unwrap().0 == e.0 {
```

Comment:
This really should be an `if let Some(ref mut last) = inner.last_mut()` or similar. However, I couldn't get it to compile, because the lifetime of the reference taken out by `last_mut` makes `inner` unmodifiable in the else branch. I think it's fixed [here](https://github.com/rust-lang/rust/issues/12033), but it'll be a bit before this makes it to stable.